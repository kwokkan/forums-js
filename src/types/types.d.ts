
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
