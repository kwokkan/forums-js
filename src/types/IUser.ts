export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
