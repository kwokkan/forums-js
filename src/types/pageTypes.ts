import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
