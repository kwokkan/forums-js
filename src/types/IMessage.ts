import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
