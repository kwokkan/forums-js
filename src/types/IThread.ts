import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
