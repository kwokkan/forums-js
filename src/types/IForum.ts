import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
