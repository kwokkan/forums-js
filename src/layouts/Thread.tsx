import React, { useState } from "react";
import { Avatar, Button, Content, FlexboxGrid, Input, Panel, PanelGroup } from "rsuite";
import { IThread } from "../types/IThread";
import { IUser } from "../types/IUser";
import { formatDate } from "../utils/dateUtils";
import { acronym } from "../utils/stringUtils";

export interface IProps {
    thread: IThread;
    user?: IUser;
    onNewComment?: (comment: string) => void;
}

export function Thread(props: IProps) {
    const [comment, setComment] = useState<string>("");

    const thread = props.thread;

    return (
        <Panel header={<h3>{thread.name}</h3>}>
            <PanelGroup>
                {thread.messages.map(x =>
                    <Panel key={x.id}>
                        <Content>
                            <FlexboxGrid>
                                <FlexboxGrid.Item>
                                    <Avatar size="lg" className="mr-3" title={x.user.name}>{acronym(x.user.name)}</Avatar>
                                </FlexboxGrid.Item>

                                <FlexboxGrid.Item>
                                    <p>{formatDate(x.created)}</p>
                                    <p>{x.content}</p>
                                </FlexboxGrid.Item>
                            </FlexboxGrid>
                        </Content>
                    </Panel>
                )}

                {props.user &&
                    <Panel>
                        <Input
                            componentClass="textarea"
                            className="mb-3"
                            rows={10}
                            placeholder="Enter comment ..."
                            style={{ resize: "both" }}
                            value={comment}
                            onChange={setComment}
                        />

                        <Button appearance="primary" onClick={() => props.onNewComment?.(comment)}>Submit</Button>
                    </Panel>
                }
            </PanelGroup>
        </Panel>
    );
}
