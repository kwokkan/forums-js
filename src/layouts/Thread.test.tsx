import React from "react";
import { create } from "react-test-renderer";
import { IThread } from "../types/IThread";
import { Thread } from "./Thread";

test("Renders with no messages", () => {
    const thread: IThread = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "Test thread",
        messages: []
    };

    const tree = create(
        <Thread thread={thread} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with messages", () => {
    const thread: IThread = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "Test thread",
        messages: [
            {
                id: 2,
                created: new Date(2020, 2, 2).getTime(),
                content: "First !!!",
                user: {
                    id: 3,
                    joinedDate: new Date(1900, 1, 1).getTime(),
                    name: "Test user"
                }
            }
        ]
    };

    const tree = create(
        <Thread thread={thread} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
