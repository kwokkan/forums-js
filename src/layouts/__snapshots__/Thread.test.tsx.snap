// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with messages 1`] = `
<div
  className="rs-panel rs-panel-default"
>
  <div
    className="rs-panel-heading"
    onClick={[Function]}
    role="rowheader"
    tabIndex={-1}
  >
    <ol
      aria-label="breadcrumbs"
      className="rs-breadcrumb"
      role="navigation"
    >
      <li
        className="rs-breadcrumb-item"
      >
        <a
          href="/"
          onClick={[Function]}
          passHref={true}
          title="Home"
        >
          <i
            className="rs-icon rs-icon-home"
          />
        </a>
      </li>
    </ol>
  </div>
  <div
    className="rs-panel-body"
  >
    <div
      className="rs-panel-group"
    >
      <div
        className="rs-panel rs-panel-default"
      >
        <div
          className="rs-panel-body"
        >
          <div
            className="rs-content"
          >
            <div
              className="rs-flex-box-grid rs-flex-box-grid-top rs-flex-box-grid-start"
            >
              <div
                className="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <div
                  className="rs-avatar mr-3 rs-avatar-lg"
                  title="Test user"
                >
                  Tu
                </div>
              </div>
              <div
                className="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <p>
                  2020-03-02T00:00:00.000Z
                </p>
                <p>
                  First !!!
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders with new message box 1`] = `
<div
  className="rs-panel rs-panel-default"
>
  <div
    className="rs-panel-heading"
    onClick={[Function]}
    role="rowheader"
    tabIndex={-1}
  >
    <ol
      aria-label="breadcrumbs"
      className="rs-breadcrumb"
      role="navigation"
    >
      <li
        className="rs-breadcrumb-item"
      >
        <a
          href="/"
          onClick={[Function]}
          passHref={true}
          title="Home"
        >
          <i
            className="rs-icon rs-icon-home"
          />
        </a>
      </li>
    </ol>
  </div>
  <div
    className="rs-panel-body"
  >
    <div
      className="rs-panel-group"
    >
      <div
        className="rs-panel rs-panel-default"
      >
        <div
          className="rs-panel-body"
        >
          <div
            className="rs-content"
          >
            <div
              className="rs-flex-box-grid rs-flex-box-grid-top rs-flex-box-grid-start"
            >
              <div
                className="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <div
                  className="rs-avatar mr-3 rs-avatar-lg"
                  title="Test user"
                >
                  Tu
                </div>
              </div>
              <div
                className="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <p>
                  2020-03-02T00:00:00.000Z
                </p>
                <p>
                  First !!!
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="rs-panel rs-panel-default"
      >
        <div
          className="rs-panel-body"
        >
          <div
            className="mb-3"
          >
            <div
              className="rs-nav rs-nav-subtle rs-nav-horizontal"
            >
              <ul>
                <li
                  className="rs-nav-item preview-strip-1 rs-nav-item-active"
                >
                  <a
                    className="rs-nav-item-content"
                    onClick={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    Text
                    <span
                      className="rs-ripple-pond"
                    >
                      <span
                        className="rs-ripple"
                        style={Object {}}
                      />
                    </span>
                  </a>
                </li>
                <li
                  className="rs-nav-item preview-strip-2"
                >
                  <a
                    className="rs-nav-item-content"
                    onClick={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    Preview
                    <span
                      className="rs-ripple-pond"
                    >
                      <span
                        className="rs-ripple"
                        style={Object {}}
                      />
                    </span>
                  </a>
                </li>
              </ul>
              <div
                className="rs-nav-waterline"
              />
            </div>
            <textarea
              className="rs-input"
              id={null}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter message..."
              rows={10}
              style={
                Object {
                  "resize": "both",
                }
              }
              type="text"
              value=""
            />
          </div>
          <button
            className="rs-btn rs-btn-primary"
            onClick={[Function]}
            type="button"
          >
            Submit
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders with no messages 1`] = `
<div
  className="rs-panel rs-panel-default"
>
  <div
    className="rs-panel-heading"
    onClick={[Function]}
    role="rowheader"
    tabIndex={-1}
  >
    <ol
      aria-label="breadcrumbs"
      className="rs-breadcrumb"
      role="navigation"
    >
      <li
        className="rs-breadcrumb-item"
      >
        <a
          href="/"
          onClick={[Function]}
          passHref={true}
          title="Home"
        >
          <i
            className="rs-icon rs-icon-home"
          />
        </a>
      </li>
    </ol>
  </div>
  <div
    className="rs-panel-body"
  >
    <div
      className="rs-panel-group"
    />
  </div>
</div>
`;
